### PR TITLE
osmocli: parse `Use` field's arguments dynamically from a message

### PR DESCRIPTION
**Pull request by pysel**
Closes: #XXX

## What is the purpose of the change

In current `osmocli` implementation, in order to create a cli command you need to create a descriptor of this command (example [here](https://github.com/osmosis-labs/osmosis/blob/main/x/concentrated-liquidity/client/cli/query.go#L44-L52)). The problem with this approach is that it is easy to forget to pass some arguments into `Use` field of descriptor, which leads to inconsistent `Usage`s  of cli commands (example [PR](https://github.com/osmosis-labs/osmosis/pull/6002) which fixes mistakenly dropped arguments).

In this PR I propose to attach arguments dynamically during command construction. It will help avoid the mentioned inconsistencies. If this gets merged, there will be no need to provide message's and query's arguments to descriptors. Ex:

```
return &osmocli.QueryDescriptor{
	Use:   "position-by-id",  // was <Use:   "position-by-id [positionID]">
        ---- snip ----
},
```

~To reviewers: right now I only implemented it for queries. If folks like the idea of this PR, I will continue to do the same for txs and migrate the repo to the new usage of osmocli~

## Testing and Verifying

Manual + added test

### CODE REVIEW

**Comment by pysel** [comment]
devbot help

**Comment by devbot-wizard** [comment]
Hi! I'm DevBot, a bot that helps with common tasks in the development process. Commands:
- `devbot fix`: Fix basic errors in the PR. (e.g. imports, changelog conflicts, go mod conflicts)
- `devbot merge base`: Merge the base branch into the PR branch.
- `devbot add changelog [misc/feat/bug/api/sec] [message]`: Add a changelog entry to the PR. (e.g. `devbot add changelog feat Added a new feature`)
    - If message is blank, defaults to PR title.
- `devbot re-pr`: Re-PR the PR, useful for external contributors where we have no edit perms.

File: x/concentrated-liquidity/tick_test.go [hunk 1]
```diff
@@ -610,7 +610,7 @@ func (s *KeeperTestSuite) TestCrossTick() {
 
 			if test.poolToGet == validPoolId {
 				s.FundAcc(s.TestAccs[0], sdk.NewCoins(DefaultCoin0, DefaultCoin1))
-				 _, err := s.clk.CreatePosition(s.Ctx, test.poolToGet, s.TestAccs[0], DefaultCoins, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
+				_, err := s.clk.CreatePosition(s.Ctx, test.poolToGet, s.TestAccs[0], DefaultCoins, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
```
Comment (pysel) [inline x/concentrated-liquidity/tick_test.go:613]:
```suggestion
				 _, err := s.clk.CreatePosition(s.Ctx, test.poolToGet, s.TestAccs[0], DefaultCoins, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
```

File: x/concentrated-liquidity/client/grpc/grpc_query.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package grpc 
+package grpc
```
Comment (pysel) [inline x/concentrated-liquidity/client/grpc/grpc_query.go:1]:
```suggestion
package grpc 
```

File: x/concentrated-liquidity/client/grpc/grpc_query.go [hunk 3]
```diff
@@ -159,4 +159,3 @@ func (q Querier) CFMMPoolIdLinkFromConcentratedPoolId(grpcCtx context.Context,
 	ctx := sdk.UnwrapSDKContext(grpcCtx)
 	return q.Q.CFMMPoolIdLinkFromConcentratedPoolId(ctx, *req)
 }
```
Comment (pysel) [inline x/concentrated-liquidity/client/grpc/grpc_query.go:161]:
```suggestion
}

```

File: osmoutils/osmocli/dynamic.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/osmosis-labs/osmosis/osmoutils"
+)
+
+type Descriptor interface {
+	GetCustomFlagOverrides() map[string]string
+	AttachToUse(str string)
+}
+
+// fields that are not provided as arguments
+var nonAttachableFields []string = []string{"sender", "pagination", "owner", "admin"}
```
Comment (pysel) [inline osmoutils/osmocli/dynamic.go:20]: to reviewers: are there some other fields which should not be passed as arguments of cli commands?

**Comment by pysel** [comment]
devbot help

**Review by p0mvn — APPROVED** [review]
utACK. Great idea!

**Comment by devbot-wizard** [comment]
Hi! I'm DevBot, a bot that helps with common tasks in the development process. Commands:
- `devbot fix`: Fix basic errors in the PR. (e.g. imports, changelog conflicts, go mod conflicts)
- `devbot merge base`: Merge the base branch into the PR branch.
- `devbot add changelog [misc/feat/bug/api/sec] [message]`: Add a changelog entry to the PR. (e.g. `devbot add changelog feat Added a new feature`)
    - If message is blank, defaults to PR title.
- `devbot re-pr`: Re-PR the PR, useful for external contributors where we have no edit perms.

**Comment by czarcas7ic** [comment]
@pysel the PR description says this is a draft, is this intended for merge? If so, could you edit the description? Thanks!

**Comment by pysel** [comment]
Hey @czarcas7ic. My bad, this is not a draft anymore

**Review by mattverse — CHANGES_REQUESTED** [review]
Hey @pysel  love this change, I think this is a great idea!

I was manually testing out things, and I noticed that arguments have trouble getting dynamically parsed methods using custom parser or custom logic. Here's an example

```
(base) matt@MacBook-Pro build % ./osmosisd q gamm estimate-swap-exact-amount-in --help
Query estimate-swap-exact-amount-in.

Example:
$ osmosisd q gamm estimate-swap-exact-amount-in 1 osm11vmx8jtggpd9u7qr0t8vxclycz85u925sazglr7 1000stake --swap-route-pool-ids=2 --swap-route-pool-ids=3

Usage:
  osmosisd query gamm estimate-swap-exact-amount-in [flags]
```

notice how it failed to dynamically place the usage. 
After we get this fixed I think we should be good to go :) Nice work!

**Comment by pysel** [comment]
devbot fix

**Comment by pysel** [comment]
hey @mattverse! weird, I just tried it on the same command, seems to be working fine:

```
ruslanakhtariev@Ruslans-MacBook-Air osmosis % osmosisd q gamm estimate-swap-exact-amount-in -h
Query estimate-swap-exact-amount-in.

Example:
$ <appd> q gamm estimate-swap-exact-amount-in 1 osm11vmx8jtggpd9u7qr0t8vxclycz85u925sazglr7 1000stake --swap-route-pool-ids=2 --swap-route-pool-ids=3

Usage:
  osmosisd query gamm estimate-swap-exact-amount-in [pool-id] [token-in] [routes] [flags]

Flags:
      --height int                   Use a specific height to query state at (this can error if the node is pruning state)
  -h, --help                         help for estimate-swap-exact-amount-in
      --node string                  <host>:<port> to Tendermint RPC interface for this chain (default "tcp://localhost:26657")
  -o, --output string                Output format (text|json) (default "text")
      --swap-route-denoms string     swap route amount
      --swap-route-pool-ids string   swap route pool id

Global Flags:
      --chain-id string     The network chain ID
      --home string         directory for config and data (default "/Users/ruslanakhtariev/.osmosisd")
      --log_format string   The logging format (json|plain) (default "plain")
      --log_level string    The logging level (trace|debug|info|warn|error|fatal|panic) (default "info")
      --trace               print out full stack trace on errors
```

could you please confirm you used an updated binary? thanks!

**Comment by mattverse** [comment]
@pysel What is the binary version you see? The current one I have is 
```
(base) matt@MacBook-Pro build % ./osmosisd version
15.8.0-333-gc63b8e409
```

**Comment by pysel** [comment]
@mattverse have you installed the binary via `go install ./...` or via some other way?

**Comment by pysel** [comment]
devbot fix

**Comment by pysel** [comment]
devbot fix changelog

**Comment by mattverse** [comment]
@pysel I use `make build` and directly use the binary file itself. Maybe it's worth looking into it

**Comment by pysel** [comment]
> @pysel I use `make build` and directly use the binary file itself. Maybe it's worth looking into it

hey @mattverse, I fixed the problem (it was outdated osmoutils version in go.mod), it should be good now.

**Comment by pysel** [comment]
btw, this osmoutils version problem took some time to be figured out. I think we might like to revisit [PR](https://github.com/osmosis-labs/osmosis/pull/4667) that was closed recently, which idea was to update versions of submodules in go.mod when anything in them changes. if we had this functionality, the mentioned problem wouldn't have occurred, wdyt?

cc: @mattverse @p0mvn @czarcas7ic

**Comment by mattverse** [comment]
I'm down to re-reviewing #4667, sorry for the delay on the review for that one

**Review by mattverse — APPROVED** [review]
LGTM, nice work 🌮

**Pull request merged**